### PR TITLE
fsm: expand fsm

### PR DESCRIPTION
**Pull request by sputn1ck**
This PR adds:
- a default observer to the FSM
- more info to the action entry and exit funcs
- an optional initial wait time for the WaitForState function

### CODE REVIEW

File: fsm/observer.go [hunk 1]
```diff
@@ -46,9 +46,43 @@ func (c *CachedObserver) GetCachedNotifications() []Notification {
 	return c.cachedNotifications.Get()
 }
 
+// WaitForStateOption is an option for the WaitForState function.
+type WaitForStateOption struct {
```
Comment (bhandras) [inline fsm/observer.go:50]: Almost, but what you want is:

```
type FSMOption interface {
    apply(*fsmOptions) 
}

type fsmOptions struct {
   initialWait time.Duration
}

type WaitForStateOption struct {
   initialWait time.Duration
}

func WithWaitForStateOption(initialWait time.Duration) FSMOption {
   return &WaitForStateOption{
         initialWait,
   }
}

func (w *WaitForStateOption) apply(o *fsmOptions) {
   o.initialWait = w.initialWait
}

```

And then below in the constructor you can pass in `opts ...FSMOption` and just call `apply()` on them.
This pattern is the same e.g. grpc uses with `DialOption` and we use the same pattern in LND a lot.

**Review by bhandras — APPROVED** [review]
LGTM, pending the option fix. 🌮

File: fsm/observer.go [hunk 2]
```diff
@@ -46,9 +46,43 @@ func (c *CachedObserver) GetCachedNotifications() []Notification {
...
+
+	var initialWait time.Duration
+	if len(opts) > 0 {
+		for _, opt := range opts {
+			initialWait = opt.initialWait
+		}
+	}
+
+	// Wait for the initial wait duration if set.
+	if initialWait > 0 {
```
Comment (hieblmi) [inline fsm/observer.go:78]: Not sure this check is needed.

File: fsm/observer.go [hunk 3]
```diff
@@ -46,9 +46,43 @@ func (c *CachedObserver) GetCachedNotifications() []Notification {
...
+// the given duration before checking the state. This is useful if the
+// function is called immediately after sending an event to the state machine
+// and the state machine needs some time to process the event.
 func (s *CachedObserver) WaitForState(ctx context.Context,
-	timeout time.Duration, state StateType) error {
+	timeout time.Duration, state StateType,
+	opts ...WaitForStateOption) error {
+
+	var initialWait time.Duration
+	if len(opts) > 0 {
```
Comment (hieblmi) [inline fsm/observer.go:71]: Can omit this check.

File: fsm/observer.go (hunk 2 above)
Comment (hieblmi) [reply fsm/observer.go:78]: Also nit on fix comment: https://github.com/lightninglabs/loop/pull/667/files#diff-28f5dffa528d8282177bcb103631cef19c6b2ae6d20abca5f559be88209469d6R129

File: fsm/observer.go [hunk 4]
```diff
@@ -46,9 +46,60 @@ func (c *CachedObserver) GetCachedNotifications() []Notification {
...
+// and the state machine needs some time to process the event.
 func (s *CachedObserver) WaitForState(ctx context.Context,
-	timeout time.Duration, state StateType) error {
+	timeout time.Duration, state StateType,
+	opts ...InitialWaitOption) error {
+
+	var options fsmOptions
+
+	for _, opt := range opts {
+		opt.apply(&options)
```
Comment (bhandras) [inline fsm/observer.go:91]: Cool! Looks good now!

**Review by bhandras — APPROVED** [review]
🚀 🚀

**Pull request merged**